### PR TITLE
feat(ta): add Tamil locale

### PR DESCRIPTION
**Pull request by tukusejssirs**
This PR add Tamil locale based on #378 and the work of @pvanthony and Friar Julian.

I marked it as WIP because it is not yet finished, however, we could still merged it into romcal before it contains all the keys localised.

### CODE REVIEW

**Comment by tukusejssirs** [comment]
@pvanthony, I made some changes to `ta.ts` you shared with us:

- I changed the key names based on #350:
   - @pvanthony, please check if I haven’t made an error;
   - all changed key names (there are 70 such keys) are listed in the description of that PR under _List of changed key names_:
      - note, however, that `mary_refuge_of_sinner` and `our_lady_of_refuge` are merged under `our_lady_refuge_of_sinners`;
      - all other key name changes are simple renames;
      - it might be a good idea to check the changed keys English values to determine if a change in `ta.ts` should also be made;
- all keys (based on `en.ts`) are now in `ta.ts`, however, keys that are kept in English (thus are not yet translated) are commented out and moved to the bottom of the `names` object:
   - all untranslated keys are copied from `en.ts` in order to update the English values when they changed;
- I moved all values on the same line as their keys;

My suggestions based on my work on romcal l10n:
- until final version is created:
   - keep the keys and their values on the same line in order to ease sorting `names` later (`prettier` helps us reformat the keys in the final version);
   - sorting `names` on its properties (key names) is not necessary until final version is created.

If you want to merge this PR before you/someone finish (-es) the translation of all keys, just tell us, we will do so.

Below are TODO lists that should be finished in order to fully translate all keys from `en.ts`, along with some (potential) issues I’ve found.

---

TODO in order to translate all keys from `en.ts` (based on #350):

- [x] localise `seasons.christmas_time.second_sunday_after_christmas`;
- [ ] localise missing keys from `names` object;
- [x] localise missing keys from `cycles` object;
- [x] are the following keys correctly localised? They contains the celebration name in English, which might be a leftover:
   - [x] `caroline_kozka_virgin`;
   - [x] `damasus_i_pope`;
   - [x] `demetrius_of_thessaloniki_martyr`;
- [x] should trailing space in key value of `adelaide_of_burgundy_empress` be removed?

TODO before merging this PR:

- [x] remove comments from `ta.ts` (untranslated keys);
- [x] sort `names` in `ta.ts`;
- [x] run `npm run prettier`;
- [x] merge after #350.

**Comment by emagnier** [comment]
> If you want to merge this PR before you/someone finish (-es) the translation of all keys, just tell us, we will do so.

I'm also in favor of this, as it will enable new uses of Romcal in Tamil and attract potential future contributors in this language 😄

**Comment by pvanthony** [comment]
On 31/8/23 03:47, Etienne Magnier wrote:
>     If you want to merge this PR before you/someone finish (-es) the
>     translation of all keys, just tell us, we will do so.
> 
> I'm also in favor of this, as it will enable new uses of Romcal in Tamil 
> and attract potential future contributors in this language 😄

Please go ahead and merge.

Sorry for the late reply.

P.V.Anthony

**Comment by tukusejssirs** [comment]
Thanks, @pvanthony, for your reply! :pray:

@emagnier, before merging, let me do some of the tasks described in [this comment of mine](https://github.com/romcal/romcal/pull/379#issuecomment-1265208374). I’ll do this after I finish PR #126 (which should be by the end of this week, at most by the end of the following week). Thanks! :pray:

**Comment by pvanthony** [comment]
Hi,

Need advice on creating the file for Sri Lankan, Sinhala.

Do I just take the tamil file, ta.ts and rename it to si.ts?

Then replace the tamil with Sinhala?

P.V.Anthony

**Comment by tukusejssirs** [comment]
> Hi, Need advice on creating the file for Sri Lankan, Sinhala. Do I just take the tamil file, ta.ts and rename it to si.ts? Then replace the tamil with Sinhala? P.V.Anthony

Hi, @pvanthony! You should always start from `en.ts` from the latest commit in the `dev` branch. :wink: You can simply copy it to a new file called `si.ts` and re-export it in `index.ts`.

Thank you in advance for your contribution! 🙏

**Comment by pvanthony** [comment]
On 31/10/23 14:18, Tukusej’s Sirs wrote:

> Hi, @pvanthony <https://github.com/pvanthony>! You should always start 
> from |en.ts| from the latest commit in the |dev| branch. 😉 You can 
> simply copy it to a new file called |si.ts| and re-export it in |index.ts|.
> 
> Thank you in advance for your contribution! 🙏

Thank you for the advice.

We will work on it. Please note that it will take some time to get it 
done completely.

By the way, was the email with the completed ta.ts file, received?

P.V.Anthony

**Comment by tukusejssirs** [comment]
> We will work on it. Please note that it will take some time to get it time to get it done completely.

That is expected. :wink:

> By the way, was the email with the completed ta.ts file, received?

Unless I have commented on it here, no. I don’t remember I have received anything like that. Could share it with us here please? 🤔

**Comment by pvanthony** [comment]
Attached is the completed Tamil, Indian language translation. I hope others can improve it if there's errors in the translation.

The work was done by Friar Julian Roy at St Mary of the Angels in Singapore.

[ta.zip](https://github.com/romcal/romcal/files/13267848/ta.zip)

**Comment by emagnier** [comment]
Thank you @pvanthony for this extensive translation work! 

I have updated this PR based on the latest version of the `ta.ts` file you shared with us (in addition to rebasing the branch on the latest codebase version). I also updated the localization keys (in a separate commit).

Just a few more small things (commented below), and then we can merge the PR 🎉.

File: lib/locales/ta.ts
Comment (emagnier) [inline lib/locales/ta.ts]: @pvanthony, can you supply a Tamil translation for this key, please?

File: lib/locales/ta.ts
Comment (emagnier) [inline lib/locales/ta.ts]: @pvanthony, can you supply a Tamil translation for this key, please?

File: lib/locales/ta.ts
Comment (emagnier) [inline lib/locales/ta.ts]: This translation was duplicated. 

@pvanthony, could you please indicate which one is more appropriate here? (FYI, it corresponds to the last Sunday of ordinary time)

File: lib/locales/ta.ts
Comment (emagnier) [inline lib/locales/ta.ts]: I found another key that would need to be reviewed or localized correctly.

File: lib/locales/ta.ts
Comment (emagnier) [inline lib/locales/ta.ts]: And this one too.

File: lib/locales/ta.ts
Comment (emagnier) [inline lib/locales/ta.ts]: And this one too.

File: lib/locales/ta.ts
Comment (emagnier) [inline lib/locales/ta.ts]: @pvanthony, if you could take a look and suggest a translation for these keys as well, that would be great!

**Comment by pvanthony** [comment]
As requested the corrections has been done.
Attached is the patch file to the original file sent by me on Nov 6th 2023.
[corrections_done_on_20240112.zip](https://github.com/romcal/romcal/files/13916712/corrections_done_on_20240112.zip)

Sorry for the delay.

**Comment by tukusejssirs** [comment]
Thanks, @pvanthony!

I’ve went quickly through the patch and I’ve some issues:

> +    catherine_of_siena_virgin_copatroness_of_europe: 'சியன்னாவின் புனித கத்தரின் - கன்னி, மறைவல்லுநர், இத்தாலி, ஐரோப்பவின் பாதுகாவலி',
> +    catherine_of_siena_virgin_copatroness_of_italy_and_europe: 'சியன்னாவின் புனித கத்தரின் - கன்னி, இத்தாலி, ஐரோப்பவின் பாதுகாவலி',

Both these keys have the same value: `Saint Catherine of Sienna - Virgin, Apostolic, Italy, Patroness of Europe`.. The patronage title should differ:
- Patroness of Europe;
- Patroness of Europe **and** Italy.

While I have kept the original translation, I suggest to verify the translation.

> -    cyril_of_jerusalem_bishop: 'அலெக்சாந்திரியா நகர்ப் புனித சிரில் - ஆயர், மறைவல்லுநர்',

~~The value is the same as for the St Cyril of Alexandria and Google Translate confirms this.~~ Actually, the value _was_ the same, my fault.

> +    cyril_the_philosopher_monk_and_methodius_of_thessaloniki_bishop_slavic_missionaries: 'புனித சிரில், புனித் மெதோடியஸ் - சிலாவிக்கின் மறை போதகர்கள்',

Could you use use `Slavic` word (`ஸ்லாவ்`) from [this article](https://www.vaticannews.va/ta/pope/news/2023-10/pope-audience-slav-saints-cyril-methodius.html) please? I believe in the key value, you used a different word (`சிலாவிக்கின்`) Or do I get wrong? :thinking:

---

I’ll try to find some time over the weekend to look at the patch thoroughly and commit changes I consider correct. I’ll keep you posted.

File: lib/locales/ta.ts
Comment (tukusejssirs) [reply lib/locales/ta.ts]: According to Google Translate, `நமது ஆண்டவராகிய இயேசு கிறிஸ்து, நித்திய தலைமைக் குரு` means `Our Lord Jesus Christ is the Eternal High Priest`, therefore, I’ll change the key name to `our_lord_jesus_christ_the_eternal_high_priest` and uncomment the other key.

File: lib/locales/ta.ts
Comment (tukusejssirs) [inline lib/locales/ta.ts]: Don't you want to add a hyphen (`-`), like you did in the value of `before_epiphany`?

```suggestion
      after_epiphany: 'திருக்காட்சிக்கு பின் - $t(weekdays:{{dow}}, capitalize)',
```

File: lib/locales/ta.ts
Comment (tukusejssirs) [inline lib/locales/ta.ts]: I believe the ranks are not properly translated. According to [this Wikipedia article](https://ta.wikipedia.org/wiki/%E0%AE%95%E0%AE%A4%E0%AF%8D%E0%AE%A4%E0%AF%8B%E0%AE%B2%E0%AE%BF%E0%AE%95%E0%AF%8D%E0%AE%95_%E0%AE%A4%E0%AE%BF%E0%AE%B0%E0%AF%81%E0%AE%9A%E0%AF%8D%E0%AE%9A%E0%AE%AA%E0%AF%88%E0%AE%AF%E0%AE%BF%E0%AE%A9%E0%AF%8D_%E0%AE%AA%E0%AF%8A%E0%AE%A4%E0%AF%81_%E0%AE%A8%E0%AE%BE%E0%AE%B3%E0%AF%8D%E0%AE%95%E0%AE%BE%E0%AE%9F%E0%AF%8D%E0%AE%9F%E0%AE%BF) (as far as I get it right), they should be as follows:

- `solemnity`: `பெருவிழா`;
- `feast`: `விழா`;
- `memorial`: `நினைவு`;
- `optional_memorial`: `விருப்ப நினைவு`;
- `weekday`: ??? `வார நாளில்` (on a weekday).

Review them please.

File: lib/locales/ta.ts
Comment (tukusejssirs) [inline lib/locales/ta.ts]: According to Google Translate, the value of both keys is (nearly, apart of `Saint,`) the same:

- `சியன்னாவின் புனித கத்தரின் - கன்னி, மறைவல்லுநர், இத்தாலி, ஐரோப்பவின் பாதுகாவலி` → `St. Catherine of Sienna - Virgin, Saint, Protectress of Italy, Europe`;
- `சியன்னாவின் புனித கத்தரின் - கன்னி, இத்தாலி, ஐரோப்பவின் பாதுகாவலி` → `Saint Catherine of Sienna - Virgin, Protectress of Italy, Europe`.

Also note that `catherine_of_siena_virgin_copatroness_of_europe` is a key used for all European countries except for Italy where there is St Catherine of Siena celebrated as _Copatroness of Europe_, while `catherine_of_siena_virgin_copatroness_of_italy_and_europe` is used in Italy only, where she is celebrated as a _Copatroness_ of both _Italy and Europe_.

Could you verify the key values, please?

File: lib/locales/ta.ts
Comment (tukusejssirs) [inline lib/locales/ta.ts]: Could you use use `Slavic` word (`ஸ்லாவ்`) from [this article](https://www.vaticannews.va/ta/pope/news/2023-10/pope-audience-slav-saints-cyril-methodius.html) please? I believe in the key value, you used a different word (`சிலாவிக்கின்`) Or do I get wrong? 🤔

**Review by tukusejssirs — COMMENTED** [review]
@pvanthony, most of the changes where only whitespace removals (you moved the values the same line as the key). I have kept the newlines, as Prettier formats the files according to our settings.

I noticed that you have added `ponntian_i_pope_and_hippolytus_of_rome_priest` key, which is incorrect. I have corrected it to `pontian_i_pope_and_hippolytus_of_rome_priest`.

I have renamed the following keys:

- `cyril_the_philosopher_monk_and_methodius_of_thessaloniki_bishop_copatrons_of_europe` → `cyril_constantine_the_philosopher_monk_and_methodius_michael_of_thessaloniki_bishop_copatrons_of_europe`;
- `cyril_the_philosopher_monk_and_methodius_of_thessaloniki_bishop_slavic_missionaries` → `cyril_constantine_the_philosopher_monk_and_methodius_michael_of_thessaloniki_bishop_slavic_missionaries_copatrons_of_europe`.

---

I have found [this Wikipedia article](https://ta.wikipedia.org/wiki/%E0%AE%95%E0%AE%A4%E0%AF%8D%E0%AE%A4%E0%AF%8B%E0%AE%B2%E0%AE%BF%E0%AE%95%E0%AF%8D%E0%AE%95_%E0%AE%A4%E0%AE%BF%E0%AE%B0%E0%AF%81%E0%AE%9A%E0%AF%8D%E0%AE%9A%E0%AE%AA%E0%AF%88%E0%AE%AF%E0%AE%BF%E0%AE%A9%E0%AF%8D_%E0%AE%AA%E0%AF%8A%E0%AE%A4%E0%AF%81_%E0%AE%A8%E0%AE%BE%E0%AE%B3%E0%AF%8D%E0%AE%95%E0%AE%BE%E0%AE%9F%E0%AF%8D%E0%AE%9F%E0%AE%BF) and in that article, there is a list of celebrations celebrated in the GRC. We/you might want to compare them with the translations we have in `ta.ts`.

I don’t say that Wikipedia article is a first-party source, however, for now, I have no Tamil source at hand. We prefer translations coming from official sources (Roman Missal, Liturgy of the Hours, Holy See decrees, etc). It would be awesome if you could provide us PDF or images of the source you use, if necessary privately.

Also, we prefer the celebration name format to be the same as used in the Roman Missal.

Note that the TODO list in [this comment of mine](https://github.com/romcal/romcal/pull/379#issuecomment-1265208374) has been fully resolved (apart of the translation of missing keys in `names` which is not a hard requirement). Also, all issues I raised in [this comment](https://github.com/romcal/romcal/pull/379#issuecomment-1888879167) were either resolved or I have created a review comment, thus that comment should be disregarded too. :wink:

---

@emagnier, I think we could merge this as is after my issues are resolved, but mark it as unofficial as I presume that some of the proper names are not official ones. Moreover, we have no source even for the celebrations which are actually celebrated in Tamil, therefore, we might want to consider this okay-ish for now.

**Comment by tukusejssirs** [comment]
I have just learnt that in Tamil, there is no concept of capital letters (there are no uppercase or lowercase letters), therefore I have removed all instances of `, capitalize`.

**Comment by pvanthony** [comment]
On 15/01/2024 03:49, Tukusej’s Sirs wrote:
> ***@***.**** commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In lib/locales/ta.ts 
> <https://github.com/romcal/romcal/pull/379#discussion_r1451792529>:
> 
>> +    our_lord_jesus_christ_king_of_the_universe: 'நமது ஆண்டவராகிய இயேசு கிறிஸ்து, நித்திய தலைமைக் குரு',
> +    // our_lord_jesus_christ_king_of_the_universe: 'இயேசு கிறிஸ்து அனைத்துலக அரசர்',
> 
> According to Google Translate, |நமது ஆண்டவராகிய இயேசு கிறிஸ்து, நித்திய 
> தலைமைக் குரு| means |Our Lord Jesus Christ is the Eternal High Priest|, 
> therefore, I’ll change the key name to 
> |our_lord_jesus_christ_the_eternal_high_priest| and uncomment the other key.

Attached is the patch to the last file we sent. Hope the corrects are ok 
now.

We do not have your copy of ta.ts. Once we have that, we will make the 
corrects on your copy of ta.ts.

P.V.Anthony

**Comment by tukusejssirs** [comment]
> We do not have your copy of ta.ts.

You can access it via GitHub either:

- [in my repository](https://github.com/tukusejssirs/romcal/blob/tamil/lib/locales/ta.ts) which is to be merged in this PR; or
- in the [PR](https://github.com/romcal/romcal/pull/379) → [`Files changed`](https://github.com/romcal/romcal/pull/379/files) tab, where you can select the [`ta.ts`](https://github.com/romcal/romcal/pull/379/files#diff-6b6111946b0e243d479cab604e84dc0ae51abcede0b7b249c3682a37941b296f) file and load the diff:
  - there are some change requets (which you can also see in this PR if you open the URL of this PR (which is always in the footer of each email notification: see `view it on GitHub` which links to the notified comment).

**Comment by pvanthony** [comment]
> On 15/01/2024 03:49, Tukusej’s Sirs wrote: ***@***.**** commented on this pull request. ------------------------------------------------------------------------ In lib/locales/ta.ts <[#379 (comment)](https://github.com/romcal/romcal/pull/379#discussion_r1451792529)>: > + our_lord_jesus_christ_king_of_the_universe: 'நமது ஆண்டவராகிய இயேசு கிறிஸ்து, நித்திய தலைமைக் குரு', + // our_lord_jesus_christ_king_of_the_universe: 'இயேசு கிறிஸ்து அனைத்துலக அரசர்', According to Google Translate, |நமது ஆண்டவராகிய இயேசு கிறிஸ்து, நித்திய தலைமைக் குரு| means |Our Lord Jesus Christ is the Eternal High Priest|, therefore, I’ll change the key name to |our_lord_jesus_christ_the_eternal_high_priest| and uncomment the other key.
> Attached is the patch to the last file we sent. Hope the corrects are ok now. We do not have your copy of ta.ts. Once we have that, we will make the corrects on your copy of ta.ts. P.V.Anthony

Looks like the attached patch did not come through the email. Here it is.
[corrections_done_on_20240122.zip](https://github.com/romcal/romcal/files/14046454/corrections_done_on_20240122.zip)

**Comment by tukusejssirs** [comment]
> Looks like the attached patch did not come through the email. Here it is. [corrections_done_on_20240122.zip](https://github.com/romcal/romcal/files/14046454/corrections_done_on_20240122.zip)

I’ll check it out next weekend.

**Comment by tukusejssirs** [comment]
I have updated this PR based on your latest patch.

---

> `cyril_the_philosopher_monk_and_methodius_of_thessaloniki_bishop_slavic_missionaries`

@pvanthony, based this key name from your latest patch file, it seems to me that you are using an older `dev` version of romcal. Some keys are changing in the `dev` version until we release the first stable `v3` version. The key names are stable unless we release a new major version and even then there needs to be a reason for it. 

Could you close my [`tamil`](https://github.com/tukusejssirs/romcal/tree/tamil) branch and test it out please? It should have everything you have posted us, therefore, I presume it should work for you. If you need instructions to test it out, just ask and I’ll give them to use.

Anyway, I suggest to use that branch from now on, until this PR is merged; then you should use `romcal@dev` which will install the latest `dev` version of romcal (which currently is `romcal@3.0.0-dev.58`).

---

@pvanthony, there are still two unresolved conversations:

- https://github.com/romcal/romcal/pull/379#discussion_r1451796253;
- https://github.com/romcal/romcal/pull/379#discussion_r1451796406.

After you resolve them, I believe we can merge this PR.

Note, however, that it would be awesome if you could add the sources where to each and every l10n key (it could be done in a separate PR). Rules:

- After the key value, Use `// src: $sourceName` comments when the key value can be found _exactly_ in the referenced document and `// based on: $sourceName` when the key value is only translated based on a particular resource.
- Provide chapter/section names after the source name/URL if required (separate each part by `, `).
- Provide page numbers at the end of the line if the source is a printed book.
- Provide page number at the end of the URL if it links to a PDF file (using `#page=$number`).
- Example format: `src: $source_title, $chapter_or_section_title, $page_number`.
- For a Roman Missal translation, the source name format should be `mr_${locale}_${publish_year}_ed${edition_number}_${differentiator}` where `${differentiator}` is optional and it should be used when multiple Roman Missal versions co-exist in a particular locale, e.g. for `en`, `es`; e.g. `mr_en_2011_ed3_us` for the English translation of the Roman Missal, ed. typ. tertia, used in the US and published in 2011).

---

Thank you for your contribution! :pray:

**Comment by emagnier** [comment]
I think we can rebase and merge as is. There are a few translations that don't seem perfect, but it's okay. Let's not let this work fall into disuse, and at least it will facilitate future uses of Romcal and contributions in this language.

**Comment by TobiTenno** [comment]
@emagnier can you spin up the ticket for the future todo's and i'll approve this PR?

**Review by TobiTenno — APPROVED** [review]

**Comment by emagnier** [comment]
I'm merging the PR. This is a long-term effort here. There's still work to be done to improve quality and validate elements with sources. For now, all the remaining todos need to be gathered into a new ticket to make tracking easier.  
But in the meantime, Romcal will already be usable in Tamil, which could potentially attract more future users and contributors in this language.

And thanks again @pvanthony for this contribution!

**Comment by emagnier** [comment]
FYI @pvanthony, the Romcal website has been updated with your translation  🎉 

https://romcal.js.org/#/calendar/ta/general-roman/2024/10

It might be easier for you to share or have certain translations reviewed by people around you who are less familiar with computing.

**Pull request merged**
